### PR TITLE
Use GString in XML code

### PR DESCRIPTION
**Pull request by nrwahl2**
I'm scrapping #2126 and starting fresh. Normally that's not advisable, but it's easier than reconciling two years worth of merge conflicts during the rebase. Plus I generally understand the code better this time around.

### CODE REVIEW

File: lib/common/acl.c [hunk 1]
```diff
@@ -85,49 +83,37 @@ create_acl(xmlNode *xml, GList *acls, enum xml_private_flags mode)
                   crm_element_name(xml), acl->xpath);
 
     } else {
-        int offset = 0;
-        char buffer[MAX_XPATH_LEN];
+        GString *xpath = g_string_sized_new(64);
```
Comment (nrwahl2) [inline lib/common/acl.c:86]: Currently allocating different amounts in different places but it's fairly arbitrary. We could define a constant in a header somewhere like 128 or 256 and just go with that.

File: lib/common/xpath.c [hunk 2]
```diff
@@ -266,38 +266,68 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
...
+    const char *id = NULL;
+    GString *xpath = NULL;
+
+    if (xml == NULL) {
+        return NULL;
+    }
+
+    xpath = pcmk__element_xpath(xml->parent);
+
+    if (TYPE(xml) == NULL) {
```
Comment (nrwahl2) [inline lib/common/xpath.c:291]: I didn't expect to have to keep this logic, but apparently the top-level XML element is nameless. Found that out the hard way.

**Comment by nrwahl2** [comment]
Not sure what I changed last night since that last time I ran `make && make install` and all the tests passed... will investigate later

File: lib/common/acl.c (hunk 1 above)
Comment (kgaillot) [reply lib/common/acl.c:86]: I think each situation is different, no need to shoehorn into one model

File: lib/common/xpath.c [hunk 3]
```diff
@@ -266,37 +266,68 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
...
-                    size_t buffer_size)
+/*!
+ * \internal
+ * \brief Get an XPath string that matches an XML element as closely as possible
+ *
+ * \param[in] xml  The XML element for which to build an XPath string
+ *
+ * \return A \p GString that matches \p xml, or \p NULL if \p xml is \p NULL.
+ *
+ * \note The caller is responsible for freeing the string.
```
Comment (kgaillot) [inline lib/common/xpath.c:277]: It's worthwhile in such notes to give the function that must be used, e.g. "using g_string_free()"

File: lib/common/xpath.c (hunk 2 above)
Comment (kgaillot) [reply lib/common/xpath.c:291]: Comments and text nodes have no name, but we probably don't ever generate xpaths for those (or have a use case for that), so maybe we should return NULL right away if xml isn't an element.

I believe all XML elements have a name so I'm not sure why the top-level element would hit this. Maybe it's actually the XML_DOCUMENT_NODE? But I would expect xml->parent and thus xpath is NULL in those cases.

File: lib/common/xpath.c [hunk 4]
```diff
@@ -266,37 +266,68 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
...
-        offset += snprintf(buffer + offset, buffer_size - offset,
-                           "/%s[@id='%s']", (const char *) xml->name, id);
-    } else if(xml->name) {
-        offset += snprintf(buffer + offset, buffer_size - offset,
-                           "/%s", (const char *) xml->name);
+    g_string_append_printf(xpath, "/%s", TYPE(xml));
+
+    id = ID(xml);
+    if (id != NULL) {
+        g_string_append_printf(xpath, "[@id='%s']", id);
```
Comment (kgaillot) [inline lib/common/xpath.c:303]: XML_ATTR_ID

**Review by kgaillot — COMMENTED** [review]
Looks good. I'm curious whether it will a have a noticeable impact on perfomance ...

File: lib/common/xpath.c (hunk 2 above)
Comment (nrwahl2) [reply lib/common/xpath.c:291]: It must have been the document node that was nameless. When I added debugging statements, I got output like `"/(null/cib/configuration/..."`.

**Comment by nrwahl2** [comment]
The current failure is from moves in `xml.c`. For example:
```
-(log_xmllib_err@xml.c:894)     error: XML Error: Entity: line 1: parser error : Start tag expected, '<' not found
-(log_xmllib_err@xml.c:894)     error: XML Error: invalidxml
-(log_xmllib_err@xml.c:894)     error: XML Error: ^
+(log_xmllib_err@xml.c:891)     error: XML Error: Entity: line 1: parser error : Start tag expected, '<' not found
+(log_xmllib_err@xml.c:891)     error: XML Error: invalidxml
+(log_xmllib_err@xml.c:891)     error: XML Error: ^
```
I'll update the expected outputs before the next update.

File: lib/common/xpath.c (hunk 4 above)
Comment (nrwahl2) [reply lib/common/xpath.c:303]: Done in next update

File: lib/common/xpath.c (hunk 3 above)
Comment (nrwahl2) [reply lib/common/xpath.c:277]: Done in next update

**Comment by kgaillot** [comment]
> The current failure is from moves in `xml.c`. For example:
> 
> ```
> -(log_xmllib_err@xml.c:894)     error: XML Error: Entity: line 1: parser error : Start tag expected, '<' not found
> -(log_xmllib_err@xml.c:894)     error: XML Error: invalidxml
> -(log_xmllib_err@xml.c:894)     error: XML Error: ^
> +(log_xmllib_err@xml.c:891)     error: XML Error: Entity: line 1: parser error : Start tag expected, '<' not found
> +(log_xmllib_err@xml.c:891)     error: XML Error: invalidxml
> +(log_xmllib_err@xml.c:891)     error: XML Error: ^
> ```
> 
> I'll update the expected outputs before the next update.

We sed out line numbers for comparison, see the last sed in cts-cli. Just need to add this function name there

File: lib/common/xpath.c (hunk 2 above)
Comment (kgaillot) [reply lib/common/xpath.c:291]: Assuming xml->parent is NULL for the document node, we should be able to just return early for anything that's not an element, then we don't need the type check

Comment (nrwahl2) [reply lib/common/xpath.c]: That seems to work. If there is somehow a node that has a parent and doesn't have a name, we're going to end up with a weird/invalid XPath no matter what. So I don't think we need to worry about that right now.

**Comment by nrwahl2** [comment]
Updated earlier commits slightly to address review and fix the test outputs, and added one new commit on top for replacing `buffer_print()`.

If this works out, I'd like to do it for the `pcmk_add_*word()` functions too, and maybe elsewhere if applicable. But before putting any more time into it, let's figure make sure performance testing results are acceptable. Two years ago in the old PR you said:

> pcmk_simtimes will be a good start, then we should also create a cluster with a really large CIB with ACLs and run some automated actions against it for a while, and compare the memory usage with and without this change. CTS with valgrind (preferably with older and newer versions of glib) will also be useful.

So I suppose I'll start there.

File: lib/common/digest.c [hunk 5]
```diff
@@ -23,22 +23,23 @@
...
 {
-    char *buffer = NULL;
-    int offset = 0, max = 0;
+    GString *buffer = g_string_sized_new(128);
 
     /* for compatibility with the old result which is used for v1 digests */
-    pcmk__buffer_add_char(&buffer, &offset, &max, ' ');
-    pcmk__xml2text(an_xml_node, 0, &buffer, &offset, &max, 0);
-    pcmk__buffer_add_char(&buffer, &offset, &max, '\n');
+
```
Comment (nrwahl2) [inline lib/common/digest.c:39]: Didn't mean to add this line. Making a note to remove it

File: lib/common/digest.c [hunk 6]
```diff
@@ -112,12 +112,13 @@ calculate_xml_digest_v2(xmlNode * source, gboolean do_filter)
          */
 
     } else {
```
Comment (nrwahl2) [inline lib/common/digest.c:114]: The `if` block concerns me. If we enter the `if (do_filter && BEST_EFFORT_STATUS)` block, then we're guaranteed to assert since `buffer` will be null. I didn't change it yet because it was already like that.

File: lib/common/xml.c [hunk 7]
```diff
@@ -1446,37 +1446,56 @@ crm_xml_escape(const char *text)
...
+        return buffer;
+    }
+
+    // Probably shouldn't happen
+    if (buffer == NULL) {
+        buffer = g_string_sized_new(128);
     }
 
-    p_name = (const char *)attr->name;
+    p_name = (const gchar *) attr->name;
```
Comment (nrwahl2) [inline lib/common/xml.c:1485]: I need to make `p_name` is a `const gchar *` variable if I'm going to do this. Probably better to just cast it in the `g_string_append_printf()` call.

Comment (nrwahl2) [reply lib/common/xml.c:1485]: Most of the casts aren't really necessary, just a safeguard in case static analysis complains later.

File: lib/common/digest.c (hunk 6 above)
Comment (kgaillot) [reply lib/common/digest.c:114]: BEST_EFFORT_STATUS is 0. We can just drop that code

File: lib/common/xml.c [hunk 8]
```diff
@@ -1947,43 +2021,47 @@ pcmk__xml2text(xmlNode *data, int options, char **buffer, int *offset,
...
+            g_string_append(buffer,
+                            (const gchar *) xmlBufContent(xml_buffer->buffer));
         }
 
 #if (PCMK__XMLDUMP_STATS - 0)
         next = time(NULL);
         if ((now + 1) < next) {
             crm_log_xml_trace(data, "Long time");
-            crm_err("xmlNodeDump() -> %dbytes took %ds", *max, next - now);
+            crm_err("xmlNodeDumpOutput() -> %lldbytes took %ds",
```
Comment (nrwahl2) [inline lib/common/xml.c:2035]: maybe add a space where one was not (before `bytes`)

File: lib/common/digest.c [hunk 9]
```diff
@@ -23,22 +23,23 @@
...
-    int offset = 0, max = 0;
+    GString *buffer = g_string_sized_new(128);
 
     /* for compatibility with the old result which is used for v1 digests */
-    pcmk__buffer_add_char(&buffer, &offset, &max, ' ');
-    pcmk__xml2text(an_xml_node, 0, &buffer, &offset, &max, 0);
-    pcmk__buffer_add_char(&buffer, &offset, &max, '\n');
+
+    g_string_append_c(buffer, ' ');
+    pcmk__xml2text(xml, 0, buffer, 0);
```
Comment (nrwahl2) [inline lib/common/digest.c:41]: Coverity thinks we leak memory here. Pretty sure it's wrong. Step 1.7 could never happen since `dump_xml_for_digest()` initializes `buffer`, so it won't be null. If it is somehow null, we've got bigger problems than a memory leak.
```
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/digest.c:41:5:
  1. alloc_fn: Storage is returned from allocation function "pcmk__xml2text".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:1983:5:
...
  1.7. path: Condition "buffer == NULL", taking true branch.
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2025:17:
  1.8. alloc_fn: Storage is returned from allocation function "g_string_sized_new".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2025:17:
  1.9. assign: Assigning: "buffer" = "g_string_sized_new(128UL)".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2027:13:
  1.10. noescape: Resource "buffer" is not freed or pointed-to in function "g_string_append".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/xml.c:2042:9:
  1.11. return_alloc: Returning allocated memory "buffer".
/srv/workspace/pacemaker-build-covscan/pacemaker-build-covscan/rhel9-coverity-x86-64/lib/common/digest.c:41:5:
  2. leaked_storage: Ignoring storage allocated by "pcmk__xml2text(xml, 0, buffer, 0)" leaks it.
```

There are four similar Coverity errors for `insert_prefix()`. I'll try assigning the return value to `buffer` every time and see if that makes Coverity happy. It's a no-op when `buffer` starts out non-null.

If that doesn't work, we could make these functions take a `GString **` argument instead of a `GString *` argument and return `void`. Or maybe even if it does work.

**Comment by nrwahl2** [comment]
`pcmk_simtimes` results from `crm_simulate cts/scheduler/xml --repeat=1000`, before and after the changes through `buffer_print()` replacement (inclusive).

[gstring_test_results.txt](https://github.com/ClusterLabs/pacemaker/files/9531553/gstring_test_results.txt)

Averages: `+0.17s  +0.77%`

Worst case: `load-stopped-loop                         97.14s vs 105.92s (+8.78s =  +9.04%)`

File: lib/common/digest.c (hunk 9 above)
Comment (nrwahl2) [reply lib/common/digest.c:41]: I went ahead and made them accept `GString **` and return `void`. Hopefully that resolves it. It's more aligned with the usual practice in Pacemaker anyway.

File: lib/common/xml.c [hunk 10]
```diff
@@ -1489,15 +1487,16 @@ pcmk__xe_log(int log_level, const char *file, const char *function, int line,
     name = crm_element_name(data);
 
     if (pcmk_is_set(options, xml_log_option_open)) {
-        char *buffer = NULL;
+        GString *buffer = g_string_sized_new(128);
```
Comment (kgaillot) [inline lib/common/xml.c:1490]: This is interesting. Some lines (e.g. a typical lrm_rsc_op) will be far longer, 512 or even bigger would be a good start for them. On the other hand most lines are likely to be smaller than 128.

I'm thinking the caller should allocate the string and pass it in, so we're only doing a single memory allocation and reusing it for each element. Probably the best approach would be to keep pcmk__xe_log() as a wrapper that allocates a large string and calls the actual function, to avoid duplication at the callers. That might even boost performance noticeably.

Regarding the initial sizes in general throughout, note that buffer_print() used a chunk size of 1024. Something between 512 and 1024 is probably about right generically, though smaller could be appropriate in some cases.

File: lib/common/xml.c [hunk 11]
```diff
@@ -1489,15 +1487,16 @@ pcmk__xe_log(int log_level, const char *file, const char *function, int line,
...
 
-        insert_prefix(options, &buffer, &offset, &max, depth);
+        insert_prefix(options, buffer, depth);
 
         if (data->type == XML_COMMENT_NODE) {
-            buffer_print(buffer, max, offset, "<!--%s-->", data->content);
+            g_string_append_printf(buffer, "<!--%s-->",
+                                   (const gchar *) data->content);
```
Comment (kgaillot) [inline lib/common/xml.c:1496]: It just dawned on me that this function should be called pcmk__xml_log() or pcmk__log_xml() since it handles comments as well as elements. It would be safer to explicitly ignore other node types.

File: lib/common/xml.c [hunk 12]
```diff
@@ -1627,7 +1624,8 @@ log_xml_changes(int log_level, const char *file, const char *function, int line,
                 const char *value = crm_element_value(data, aname);
                 flags = prefix_del;
                 do_crm_log_alias(log_level, file, function, line,
-                                 "%s %s @%s=%s", flags, spaces, aname, value);
+                                 "%s %s @%s=%s", flags,
+                                 (const char *) spaces->str, aname, value);
```
Comment (kgaillot) [inline lib/common/xml.c:1628]: Would it be straightforward to get rid of `spaces` altogether, and just use the width approach?

**Review by kgaillot — COMMENTED** [review]
> I started working on similar changes elsewhere, but I want to see where we land on this first and find out what we consider "a noticeable impact on performance." The others will all be self-contained and could even go in their own PRs. What level of degradation do you think is a threshold for "no longer acceptable"?

We don't really have a good handle on that. Actually I'm less worried about the impact of any one change vs 1-3% creep each release adding up over time.

Looking at a single change, I think +5% average would be a red flag. Given that we're running 1000 tests, the actual changes tend to be far below what would matter in the real world, as far as the scheduler running a single time. Also if the changes are in hundredths of a second across 1000 tests I think we can just ignore those (which is what -s 0.1 does). Even subsecond (so <1ms per run) doesn't really matter but it can be interesting to see those impacts.

File: lib/common/xpath.c [hunk 13]
```diff
@@ -266,37 +266,65 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
...
 {
-    const char *id = ID(xml);
+    const char *id = NULL;
+    GString *xpath = NULL;
 
-    if ((offset == 0) && (xml->parent != NULL)) {
-        offset = pcmk__element_xpath(xml->parent, buffer, offset, buffer_size);
+    // The document node has no parent and no name
+    if (xml->parent == NULL) {
+        return NULL;
```
Comment (nrwahl2) [inline lib/common/xpath.c:288]: This needs a little tune-up... if we pass in the document node, we get `NULL` back.

File: lib/common/xpath.c [hunk 14]
```diff
@@ -266,37 +266,65 @@ get_xpath_object(const char *xpath, xmlNode * xml_obj, int error_level)
...
+    GString *g_path = pcmk__element_xpath(xml);
 
-    if (pcmk__element_xpath(xml, buffer, offset, sizeof(buffer)) > 0) {
-        return strdup(buffer);
+    if (g_path == NULL) {
+        return NULL;
     }
-    return NULL;
+
+    path = strdup((const char *) g_path->str);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xpath.c:325]: Will do. FWIW I don't think `g_path->str` can be null if `g_path != NULL` (unless you manually go in and set `g_path->str` to `NULL` instead of using GLib functions).

File: lib/common/xml.c [hunk 15]
```diff
@@ -62,39 +60,31 @@ pcmk__tracking_xml_changes(xmlNode *xml, bool lazy)
...
 
-        if ((*buffer) == NULL || spaces >= ((*max) - (*offset))) {
-            (*max) = QB_MAX(CHUNK_SIZE, (*max) * 2);
-            (*buffer) = pcmk__realloc((*buffer), (*max));
+        for (int lpc = 0; lpc < spaces; lpc++) {
+            /* It's either do this (one at a time), or define a long string full
+             * of spaces and call g_string_append_len() -- which might not be
+             * better anyway.
+             */
+            g_string_append_c(buffer, ' ');
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c:86]: That's a good idea

File: lib/common/xml.c [hunk 16]
```diff
@@ -1446,14 +1436,23 @@ crm_xml_escape(const char *text)
...
 
+/*!
+ * \internal
+ * \brief Append an XML attribute to a buffer
+ *
+ * \param[in]     attr     Attribute to append
+ * \param[in]     options  Group of \p xml_log_options flags
+ * \param[in,out] buffer   Where to append the content (must not be \p NULL)
+ */
 static inline void
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c:1447]: yeah me either, will de-inline it

File: lib/common/xml.c (hunk 12 above)
Comment (nrwahl2) [reply lib/common/xml.c:1628]: More or less yes. In this context, `insert_prefix()` is useful mostly as a convenience wrapper for checking whether `log_xml_option_formatted` is set. We don't really need to append to a buffer here.

File: lib/common/xml.c (hunk 11 above)
Comment (nrwahl2) [reply lib/common/xml.c:1496]: May be a tad confusing relative to `crm_log_xml.*()` but more accurate

File: lib/common/digest.c [hunk 17]
```diff
@@ -68,14 +68,14 @@ calculate_xml_digest_v1(xmlNode * input, gboolean sort, gboolean ignored)
     }
 
     buffer = dump_xml_for_digest(input);
-    CRM_CHECK(buffer != NULL && strlen(buffer) > 0, free_xml(copy);
-              free(buffer);
+    CRM_CHECK((buffer != NULL) && (buffer->len > 0), free_xml(copy);
+              g_string_free(buffer, TRUE);
               return NULL);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/digest.c:73]: `free(buffer)` probably wasn't a great move in the first place, if `buffer` might be `NULL`.

File: lib/common/xml.c (hunk 10 above)
Comment (nrwahl2) [reply lib/common/xml.c:1490]: > Probably the best approach would be to keep pcmk__xe_log() as a wrapper that allocates a large string and calls the actual function, to avoid duplication at the callers.

That sounds like a good idea. Allocate one buffer in advance, and truncate where we're currently allocating. Truncation is lightweight for GStrings.

> Regarding the initial sizes in general throughout, note that buffer_print() used a chunk size of 1024. Something between 512 and 1024 is probably about right generically

True. The extra overhead for the initial allocation is probably minimal at that scale, and allocating a larger amount reduces the likelihood of needing to reallocate. I'll try using 4096 in `digest.c` (since it looks like we expect to handle the whole tree) and 1024 elsewhere.

**Comment by nrwahl2** [comment]
Surprisingly, none of the changes under consideration (e.g., initial buffer size, allocating a buffer just once for `pcmk__xe_log()`, using a width-based approach with `insert_prefix()`, eliminating one of the `insert_prefix()` calls and its associated `GString`, etc.) seem to make a consistent and meaningful difference in performance. (Some of the test results like `year-2038` vary between runs, even with 1000 repeats -- about 2 seconds' variation for that one). I went so far as implementing them one at a time, testing in between.

I'll update anyway, since some of them weren't intended to impact performance and the others don't hurt.

File: lib/common/xml.c [hunk 18]
```diff
@@ -1682,6 +1684,9 @@ log_data_element(int log_level, const char *file, const char *function,
 
     char *prefix_m = NULL;
 
+    // @COMPAT: depth should become unsigned at a compatibility break
+    CRM_ASSERT(depth >= 0);
```
Comment (nrwahl2) [inline lib/common/xml.c:1705]: This is effectively the same thing as changing the interface to `unsigned depth`, which I was trying to avoid for this public API function, so I wonder if I can do this either. On the one hand, if someone is passing a negative depth, that's a bug in their code. On the other, I think it would have "worked" before, in some form or fashion.

File: lib/common/xpath.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/xpath.c:333]: I mean assert the strdup() result is non-NULL. We can assume g_path->str is non-NULL

File: lib/common/digest.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/digest.c:73]: free(NULL) is perfectly fine, which makes life easier. All our custom object free functions accept NULL as well. However that can't be assumed for glib and libxml2.

File: lib/common/xml.c (hunk 18 above)
Comment (kgaillot) [reply lib/common/xml.c:1705]: The compatibility break is more about the function signature. Significant behavioral changes can be considered compatibility breaks as well, but I don't think this rises to that level, it's just improved error-checking.

File: lib/common/patchset.c [hunk 19]
```diff
@@ -539,13 +539,9 @@ xml_log_patchset(uint8_t log_level, const char *function, xmlNode *patchset)
 
             } else if (strcmp(op, "modify") == 0) {
                 xmlNode *clist = first_named_child(change, XML_DIFF_LIST);
-                char buffer_set[PCMK__BUFFER_SIZE];
-                char buffer_unset[PCMK__BUFFER_SIZE];
-                int o_set = 0;
-                int o_unset = 0;
+                GString *buffer_set = g_string_sized_new(256);
+                GString *buffer_unset = g_string_sized_new(256);
```
Comment (kgaillot) [inline lib/common/patchset.c:543]: It would be more efficient to take these variables out of the loop, allocate them here if they're null otherwise truncate them, and free them after the loop.

**Review by kgaillot — COMMENTED** [review]
looks good

File: cts/cts-cli.in [hunk 20]
```diff
@@ -2419,6 +2419,7 @@ for t in $tests; do
         -e 's/.*\(unpack_.*\)@.*\.c:[0-9][0-9]*)/\1/g' \
         -e 's/.*\(update_validation\)@.*\.c:[0-9][0-9]*)/\1/g' \
         -e 's/.*\(apply_upgrade\)@.*\.c:[0-9][0-9]*)/\1/g' \
+        -e 's/.*\(log_xmllib_err\)@.*\.c:[0-9][0-9]*)/\1/g' \
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply cts/cts-cli.in:2422]: It needs to be removed :) Taken care of here: https://github.com/ClusterLabs/pacemaker/commit/550e22c

File: lib/common/xpath.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xpath.c:333]: will do

File: lib/common/xpath.c (hunk 13 above)
Comment (nrwahl2) [reply lib/common/xpath.c:287]: GitHub doesn't show my comment as outdated but it's referring to a previous version of this function. I think the function as written now should work fine for any node.

If we want, I can definitely still restrict it to non-document in the doxygen and modify the code accordingly. It would consolidate two `g_string_*()` calls into one when (and only when) there's an ID.

Comment (kgaillot) [reply lib/common/xpath.c:287]: Not worth changing

**Comment by nrwahl2** [comment]
Not having much luck so far in getting to whatever `buffer_print() -> GString` change had good performance. I grabbed what I think is the original commit that I had saved in another branch, and it still performs poorly (average =~ +4.64%) compared to the initial tests (average < +1%) (see https://github.com/ClusterLabs/pacemaker/pull/2857#issuecomment-1241391959).

Frustrating... current results are on the border of what **might** be allowable, but if we could get a 1% or less increase, that would be wonderful

**Comment by nrwahl2** [comment]
I might be making some progress. I got things back down to good levels and am once again building back up from there, one small change at a time.

The "deinline `dump_xml_attr()` change, which I'm testing right now, seems to cause a significant spike (from about +1% to about +1.6%). So I'll probably drop that change, and I'll probably experiment with inlining some other small static functions in `xml.c`.

File: lib/common/xml.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c]: The "deinline `dump_xml_attr()`" change seems to cause a significant spike (from about +1% to about +1.6%). So I'll drop that change, and I may experiment with inlining some other small static functions in `xml.c`.

I did some cursory research on when gcc automatically inlines functions. For the most part, how it decides isn't advertised. It uses heuristics, one of which is the impact on code size. I think it'll normally inline static functions that are called only once; this one is called more than once.

File: lib/common/xml.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/xml.c]: It turns out `g_string_append_printf()` performs significantly worse than the loop of `g_string_append_c()` calls, presumably due to the overhead of dealing with the format string and va args. `g_string_append_c()` requires more calls (though compiler optimization might eliminate some of them), but the function itself is pretty lightweight.

**Comment by nrwahl2** [comment]
Converted to draft. I'm running tests overnight to confirm final perf results and make sure nothing weird in regression.

File: lib/common/acl.c [hunk 21]
```diff
@@ -85,27 +85,20 @@ create_acl(xmlNode *xml, GList *acls, enum xml_private_flags mode)
     } else {
         GString *buf = g_string_sized_new(128);
```
Comment (nrwahl2) [inline lib/common/acl.c:86]: You'd think allocating the buffer just once (if and when it's needed) instead of on every `create_acl()` call would be more efficient. Empirically it either made no difference or performed slightly worse on `crm_simulate --profile` and `cts-cli -t acls` tests, compared to the current approach.

**Comment by nrwahl2** [comment]
Okay, it's happening again... I made every change one-by-one in a temp branch. After every change, I did a complete fresh rebuild: uninstall, clean, make, install. I ran a full 1000-repeat `crm_simulate --profile cts/scheduler/xml` after each change and collected the results, and the final ones were really solid (< +0.50%).

I destroyed my `nrwahl2-T152` branch, re-created it based on current main, and cherry-picked all my commits from the temp branch onto `nrwahl2-T152`. Uninstalled, cleaned, made, installed, and ran the `crm_simulate` test again. Absolute crap results (> +5%), using the exact same commits. There is some definite weirdness that I can't explain.

I now feel fairly confident that the changes are okay. I have no idea why *the act of working from pulled-down or cherry-picked commits instead of commits made directly onto the current branch* somehow degrades performance.

File: lib/common/acl.c (hunk 21 above)
Comment (kgaillot) [reply lib/common/acl.c:86]: It's possible glib is significantly less efficient than the standard functions, and/or it's sensitive to the initial size

File: lib/common/xml.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/xml.c]: good to know

**Comment by kgaillot** [comment]
> I have no idea why the act of working from pulled-down or cherry-picked commits instead of commits made directly onto the current branch somehow degrades performance.

Yeah that's concerning. I wonder if doing so many runs one-by-one seeded the disk cache better. Have you tried running the cherry-picked version repeatedly to see if the performance starts to improve at some point?

We do need to figure out how to make the timings more consistent from run to run. A quick search suggests `sync && echo 1 > /proc/sys/vm/drop_caches` might be a good starting point. Alternatively, putting the checkout on a separate file system and unmounting/remounting it before each test might clear the caches. Another alternative might be https://github.com/hoytech/vmtouch which can lock certain files in cache.

**Review by kgaillot — APPROVED** [review]
Looks good, let me know when you're ready to merge

File: lib/common/xml.c (hunk 10 above)
Comment (nrwahl2) [reply lib/common/xml.c]: I thought I bumped the buffers in `pcmk__xml_log()` to 1024 too. I won't change it now since a later commit replaces both of these calls with a single one in a wrapper.

File: lib/common/acl.c [hunk 22]
```diff
@@ -262,14 +241,27 @@ pcmk__apply_acl(xmlNode *xml)
...
 
             p = match->_private;
-            crm_trace("Applying %s ACL to %s matched by %s",
-                      acl_to_text(acl->mode), path, acl->xpath);
             pcmk__set_xml_flags(p, acl->mode);
-            free(path);
+
+            /* Build a GString only if tracing is enabled.
+             * Can't use pcmk__log_else() because the else_action would be
+             * continue.
```
Comment (nrwahl2) [inline lib/common/acl.c:252]: We might want to create a `pcmk__log_if()` or `pcmk__log_then()` macro that acts like a flipped `pcmk__log_else()`: run this block only if we WOULD log a message. Would be a separate PR at this point if we did

**Comment by nrwahl2** [comment]
> Have you tried running the cherry-picked version repeatedly to see if the performance starts to improve at some point?

Not really. Six straight runs of 1000 repeats each on just the `year-2038.xml` CIB:
```
Averages: +2.07s  +4.43%
Averages: +1.53s  +3.27%
Averages: +1.82s  +3.89%
Averages: +1.96s  +4.19%
Averages: +1.69s  +3.62%
Averages: +2.24s  +4.79%
```

Code comparison on relevant files to the temp branch where I achieved all the positive test results (separated by full rebuilds):
```
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test lib/common/xml.c
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test lib/common/acl.c
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test lib/common/crmcommon_private.h
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test lib/common/digest.c
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test lib/common/xpath.c
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test lib/common/patchset.c
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test include/crm/common/xml.h
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test include/crm/common/xml_compat.h
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test include/crm/common/xml_internal.h
[root@boomstick pacemaker]# git diff nrwahl2-T152..nrwahl2-T152_test daemons/controld/controld_te_utils.c
diff --git a/daemons/controld/controld_te_utils.c b/daemons/controld/controld_te_utils.c
index 8e86ee98d..1aeb7acd7 100644
--- a/daemons/controld/controld_te_utils.c
+++ b/daemons/controld/controld_te_utils.c
[const differences due to last rebase-on-main time]
```

> We do need to figure out how to make the timings more consistent from run to run.

The timings are fairly consistent, depending on how you define it. The same code tended to produce results < 0.2 percentage points apart during the good runs.

Three runs of `year-2038.xml` on the cherry-picked branch with syncs in between perform better on average than without syncs, but I chalk that up to random variation.
```
Averages: +1.70s  +3.64%
Averages: +1.34s  +2.87%
Averages: +1.54s  +3.30%
```

---

When I flip back over to the temp branch and do a complete rebuild, things aren't as nice as they were (< +0.5%) before moving to the cherry-picked branch but they're still within tolerable limits:
```
Averages: +0.04s  +1.13%
```

And since I provided only `year-2038` results on the cherry-picked branch (though full results were trending around +6%), here's the `year-2038` result from the temp branch:
```
year-2038                                 46.73s vs  47.23s (+0.50s =  +1.07%)
```

---

> let me know when you're ready to merge

The extra `cts-cli.in` change is still there -- I removed it but that must have been part of an aborted rebase or something. I'm about to push that, and then when CI finishes we can merge.
